### PR TITLE
Update rust-web3 to ~ v0.15.0

### PR DESCRIPTION
**Pull request by lutter**
There are two things that I'd like to get more input on:

* in `chain/ethereum/src/trigger.rs` we need to deal with `Transaction.from` now being an `Option`, but there's no good way to handle `None`; what was the previous behavior when the raw transaction had no `from`? I couldn't figure that out from the `rust-web3` sources. How should we handle that now? I made that case panic, but that's probably not the best solution.
* in `chain/ethereum/src/transport.rs` we do away with setting the maximum parallelism for HTTP requests. Is that ok?

### CODE REVIEW

File: chain/ethereum/src/trigger.rs [hunk 1]
```diff
@@ -359,15 +363,19 @@ pub struct EthereumTransactionData {
 
 impl From<&'_ Transaction> for EthereumTransactionData {
     fn from(tx: &Transaction) -> EthereumTransactionData {
+        // Old versions of web3 (before commit 11277e07) always had a value
+        // for `from`; newer versions make this an option, but it's not
+        // clear what the value should be if this is ever `None`
+        let from = tx.from.expect("tx.from must have a value");
```
[the remark this answers is not in the document]
Comment (lutter) [reply chain/ethereum/src/trigger.rs:369]: updated the comment

File: chain/ethereum/src/ethereum_adapter.rs [hunk 2]
```diff
@@ -490,121 +480,122 @@ impl EthereumAdapter {
...
-
-                        // 0xfd is REVERT, but on some contracts, and only on older blocks,
-                        // this happens. Makes sense to consider it a revert as well.
-                        const PARITY_BAD_INSTRUCTION_FD: &str = "Bad instruction fd";
-
-                        const PARITY_BAD_JUMP_PREFIX: &str = "Bad jump";
-                        const PARITY_STACK_LIMIT_PREFIX: &str = "Out of stack";
-
-                        // See f0af4ab0-6b7c-4b68-9141-5b79346a5f61.
-                        const PARITY_OUT_OF_GAS: &str = "Out of gas";
```
[the remark this answers is not in the document]
Comment (lutter) [reply chain/ethereum/src/ethereum_adapter.rs:520]: Restored the out of gas check

File: chain/ethereum/src/ethereum_adapter.rs [hunk 3]
```diff
@@ -490,121 +480,122 @@ impl EthereumAdapter {
...
 
-                                    // The VM execution error was not identified as a revert.
-                                    _ => Err(EthereumContractCallError::Web3Error(
-                                        web3::Error::Rpc(rpc_error.clone()),
-                                    )),
+                        // Check for Geth revert.
+                        Err(web3::Error::Rpc(rpc_error))
+                            if GETH_EXECUTION_ERRORS
+                                .iter()
+                                .any(|e| rpc_error.message.contains(e)) =>
```
[the remark this answers is not in the document]
Comment (lutter) [reply chain/ethereum/src/ethereum_adapter.rs:549]: Fixed

File: chain/ethereum/src/ethereum_adapter.rs [hunk 4]
```diff
@@ -838,7 +832,14 @@ impl EthereumAdapter {
             retry("chain_id RPC call", &logger)
                 .no_limit()
                 .timeout_secs(*JSON_RPC_TIMEOUT)
-                .run(move || web3.eth().chain_id().from_err::<Error>().compat())
+                .run(move || {
+                    let web3 = web3.cheap_clone();
+                    async move {
+                        {
+                            web3.eth().chain_id().await
+                        }
```
[the remark this answers is not in the document]
Comment (lutter) [reply chain/ethereum/src/ethereum_adapter.rs:840]: Removed

File: chain/ethereum/src/ethereum_adapter.rs [hunk 5]
```diff
@@ -862,70 +863,62 @@ impl EthereumAdapterTrait for EthereumAdapter {
         let net_version_future = retry("net_version RPC call", &logger)
             .no_limit()
             .timeout_secs(20)
-            .run(move || web3.net().version().from_err().compat())
-            .boxed()
-            .compat();
+            .run(move || {
+                let web3 = web3.cheap_clone();
+                async move { web3.net().version().await.with_context(|| "boom") }
```
[the remark this answers is not in the document]
Comment (lutter) [reply chain/ethereum/src/ethereum_adapter.rs:868]: Less :boom:

**Comment by lutter** [comment]
Addressed review comments and rebased against latest `master`. I also pushed a branch `lutter/web3-old` that is the original branch for this PR + the fixes from the review. Thanks for the thorough review!

File: chain/ethereum/src/ethereum_adapter.rs [hunk 6]
```diff
@@ -117,19 +118,6 @@ lazy_static! {
...
-// subgraphs come across other errors. See
-// https://github.com/ethereum/go-ethereum/blob/dfeb2f7e8001aef1005a8d5e1605bae1de0b4f12/core/vm/errors.go#L25-L38
-const GETH_ETH_CALL_ERRORS: &[&str] = &[
-    "execution reverted",
-    "invalid jump destination",
-    "invalid opcode",
-    // Ethereum says 1024 is the stack sizes limit, so this is deterministic.
-    "stack limit reached 1024",
-    // See f0af4ab0-6b7c-4b68-9141-5b79346a5f61 for why the gas limit is considered deterministic.
-    "out of gas",
```
Comment (leoyvens) [inline chain/ethereum/src/ethereum_adapter.rs:130]: This line (and the comment) was lost.

File: chain/ethereum/src/ethereum_adapter.rs [hunk 7]
```diff
@@ -492,121 +481,126 @@ impl EthereumAdapter {
...
-                        const GANACHE_VM_EXECUTION_ERROR: i64 = -32000;
-                        const GANACHE_REVERT_MESSAGE: &str =
-                            "VM Exception while processing transaction: revert";
-                        const PARITY_VM_EXECUTION_ERROR: i64 = -32015;
-                        const PARITY_REVERT_PREFIX: &str = "Reverted 0x";
-
-                        let mut geth_execution_errors = GETH_ETH_CALL_ERRORS
-                            .iter()
-                            .map(|s| *s)
-                            .chain(GETH_ETH_CALL_ERRORS_ENV.iter().map(|s| s.as_str()));
```
Comment (leoyvens) [inline chain/ethereum/src/ethereum_adapter.rs:533]: The logic of chaining with `GETH_ETH_CALL_ERRORS_ENV` was lost.

**Review by leoyvens — CHANGES_REQUESTED** [review]

**Comment by leoyvens** [comment]
Apologies for the lack of unit tests on this stuff but some of the revert detection logic wasn't fully restored.

**Comment by lutter** [comment]
> Apologies for the lack of unit tests on this stuff but some of the revert detection logic wasn't fully restored.

Sorry this turned into such a game of whack-a-mole; should be addressed now.

I think the issue isn't the lack of unit tests, but more that there's no separation between the error detection logic and the logic controlling the overall flow of using `eth_call`. One day, I'll factor the former into some sort of helper ;)

**Comment by lutter** [comment]
Rebased to latest master

**Review by leoyvens — APPROVED** [review]

**Comment by lutter** [comment]
I made a git mistake when merging this; merged this with commit 39d9c3a5

**Comment by lutter** [comment]
Before merging, I had run this in the integration cluster with `integer` at 2345ddf5. Only differences were in the bahvior of subgraph failure, where this branch reported one block later than the code in `integer` which is expected

**Pull request closed without merging**